Handle addressing modes for HW intrinsics

CarolEidt opened the pull request:
Contribute to #19550
Fix #19521

mikedn commented on src/jit/codegenlinear.cpp:
```diff
@@ -1333,6 +1333,25 @@ void CodeGen::genConsumeRegs(GenTree* tree)
             genUpdateLife(tree);
         }
+#ifdef FEATURE_HW_INTRINSICS
+        else if (tree->OperIs(GT_HWIntrinsic))
```
`GT_HWIntrinsic` is already handled a few lines below.

tannergooding replied:
Did this get addressed? I don't see the matching removal

CarolEidt replied:
It's a bit hard to see, with the way the diffs are shown, but if you look below, this line was deleted:
```
        else if (tree->OperIsHWIntrinsic())
```
and the `genConsumeReg` below it is actually the one for the `OperIsInitVal` case.

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -109,32 +109,50 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    genConsumeAddress(op1);
+                    // Until we improve the handling of addressing modes in the emitter, we'll create a
+                    // temporary GT_IND to generate code with.
+                    GenTreeIndir load = indirForm(node->TypeGet(), op1);
```
Meh, and I just got rid of the last usage of `indirForm` in XARCH codegen in a recent PR :disappointed:  . I'll clean this up once the 3.0 release storm passes.

mikedn commented on src/jit/emitxarch.cpp:
```diff
@@ -4088,6 +4088,119 @@ void emitter::emitIns_R_A_I(instruction ins, emitAttr attr, regNumber reg1, GenT
+//    rmOp      -- The second operand, which may be a memory node or a node producing a register
+//    ival      -- The immediate operand
+//
+void emitter::emitIns_R_RM_I(instruction ins, emitAttr attr, regNumber reg1, GenTree* rmOp, int ival)
```
I wouldn't put this in the emitter. Ideally the emitter interface should not deal with `GenTree`, only instructions, registers and immediates/displacements.

tannergooding replied:
You could probably just update `genHWIntrinsic_R_RM_I()`: https://github.com/dotnet/coreclr/blob/407cc7d69c1db869e2f4418e180f51394fa3c73d/src/jit/hwintrinsiccodegenxarch.cpp#L528

tannergooding replied:
Hmm, it actually looks like most of this logic was pulled from the previous impl of `genHWIntrinsic_R_RM_I`

tannergooding replied:
I think most of the `genHWIntrinsic_R_...` methods are probably more "general-purpose" than HWIntrinsic specific.

That is, outside some asserts and the explicit code path for `containedOp->OperIsHWIntrinsic()`, they are just taking a node and determining which emitter call to make (`R`, `AR`, `C`, `A`, `S`, etc).

It might be worth refactoring all of these to be more generally usable?

mikedn replied:
The various "RM" handling functions and the relevant emitter interface functions are pretty messed up as is. It will take  a significant effort to address all that and it certainly doesn't belong in this PR. For now juts avoiding adding more stuff to the emitter interface should be enough. It already suffers from duplication, bad function and parameter names etc.

tannergooding replied:
> It will take a significant effort to address all that and it certainly doesn't belong in this PR.

Completely right, I was more asking as a potential future work item.

CarolEidt replied:
As it turns out, extracting the HWIntrinsic-dependencies from this was a bit problematic, so although I moved to to `CodeGen::inst_RV_TT_IV()`, it is still under `#ifdef FEATURE_HW_INTRINSICS`.

tannergooding commented on src/jit/codegenlinear.cpp:
```diff
@@ -1389,6 +1403,45 @@ void CodeGen::genConsumeOperands(GenTreeOp* tree)
+        return;
+    }
+    if (op1->OperIs(GT_LIST))
+    {
```
nit: `assert(node->gtGetOp2() == nullptr)` when `op1` is a list?

tannergooding commented on src/jit/codegenlinear.cpp:
```diff
@@ -1389,6 +1403,45 @@ void CodeGen::genConsumeOperands(GenTreeOp* tree)
+        assert(node->gtGetOp2() == nullptr);
+        return;
+    }
+    if (op1->OperIs(GT_LIST))
```
nit: `OperIsList()`?

mikedn replied:
I wouldn't bother too much with this. Hopefully we'll just get rid of `GT_LIST`. I have ~75% of it done, just waiting for intrinsics to finally stabilize so I don't have to keep fixing conflicts.

tannergooding commented on src/jit/codegenlinear.cpp:
```diff
@@ -1389,6 +1403,45 @@ void CodeGen::genConsumeOperands(GenTreeOp* tree)
+//    None.
+//
+
+void CodeGen::genConsumeHWIntrinsicOperands(GenTreeHWIntrinsic* node)
```
Might be nice to centralize the asserting of `numArgs` to number of node operands here as well?

tannergooding commented on src/jit/gentree.cpp:
```diff
@@ -3510,6 +3510,28 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
+                {
+                    HWIntrinsicCategory category =
+                        HWIntrinsicInfo::lookupCategory(tree->AsHWIntrinsic()->gtHWIntrinsicId);
+                    if ((category == HW_Category_MemoryLoad) || (category == HW_Category_MemoryStore))
```
nit: `OperIsMemoryLoadOrStore()`?

CarolEidt replied:
Will do

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -153,25 +171,30 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     assert((ival >= 0) && (ival <= 127));
                     genHWIntrinsic_R_R_RM_I(node, ins, ival);
                 }
                 else if (category == HW_Category_MemoryLoad)
```
Why isn't this case `genConsumeAddress` like the other loads?

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1803,16 +1732,13 @@ void CodeGen::genSSE41Intrinsic(GenTreeHWIntrinsic* node)
-        op1Reg = op1->gtRegNum;
-        genConsumeOperands(node);
-    }
+    genConsumeHWIntrinsicOperands(node);
```
Shouldn't there be some additional removals of `genConsumeReg` from the below (for this and similar edits)?

CarolEidt replied:
I didn't transform all the methods in this file to use `genConsumeHWIntrinsicOperands` - only the ones where it was specifically beneficial to abstract out the possibility that they might be contained memory ops/addresses. That may be a useful future change, but I wouldn't necessarily do that with this PR.

CarolEidt commented:
test OSX10.12 x64 Checked Innerloop Build and Test
test Windows_NT arm Cross Checked Innerloop Build and Test

CarolEidt commented:
@dotnet/jit-contrib PTAL

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -109,32 +109,50 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    // Until we improve the handling of addressing modes in the emitter, we'll create a
+                    // temporary GT_IND to generate code with.
+                    GenTreeIndir load = indirForm(node->TypeGet(), op1);
+                    emit->emitInsLoadInd(ins, simdSize, node->gtRegNum, &load);
```
The old code was using `emitIns_R_AR`, that one has a `Is4ByteSSEInstruction` check that increases the instruction size by 2. `emitInsLoadInd` does not seem to have such a check.

CarolEidt replied:
Right - I took a quick look at that when you made that comment earlier, and I wasn't sure it was actually required. I need to take another look.

mikedn replied:
That 4 byte check appeared in #16249 and was updated in #21528. It's not clear to me why that was done in `emitIns` functions instead of `emitInsSize` functions.

CarolEidt replied:
So far I believe that this 2-byte increment should actually be 1 byte. That said, there is definitely inconsistency in how the sizes are determined, and the code allows the size to be over-estimated, so those errors go undetected. I'm trying to decide whether to make a more general cleanup here or just go with the 1 byte increment for this case.

tannergooding replied:
I thought we had a work item tracking the cleanup of `emitInsSize`, but I can't seem to find it.

From what I remember, these one offs were done because it was non-trivial to update the existing `emitInsSize` checks due to other complications at the time. I think it is `+2` to handle the non-VEX encoded case but only needs to be `+1` for the VEX encoding.

CarolEidt replied:
I believe it's the case that for these addressing-mode instructions, the SSE encodings require one additional bit (and the VEX encodings don't require an additional bytes). I've run all the hw intrinsic tests without AVX & AVX2 with the most recent set of changes.
The size analysis is so convoluted that I am mostly relying on empirical results (i.e. testing).

CarolEidt replied:
I added a note to https://github.com/dotnet/coreclr/issues/23006 that this should also be addressed.

CarolEidt commented on src/jit/instr.cpp:
```diff
@@ -1039,7 +1039,120 @@ void CodeGen::inst_RV_RV_IV(instruction ins, emitAttr size, regNumber reg1, regN
+    else
+    {
+        regNumber rmOpReg = rmOp->gtRegNum;
+        getEmitter()->emitIns_SIMD_R_R_I(ins, attr, reg1, rmOpReg, ival);
```
This isn't really SIMD-specific (and should probably be moved), but it's the method that will do a copy to the target if needed for the RMW form.

CarolEidt commented:
@dotnet-bot test Ubuntu x64 Checked CoreFX Test
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Formatting

CarolEidt commented:
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests

CarolEidt commented:
@dotnet/jit-contrib @tannergooding @fiigii PTAL

tannergooding commented on src/jit/codegenlinear.cpp:
```diff
@@ -1389,6 +1399,55 @@ void CodeGen::genConsumeOperands(GenTreeOp* tree)
+    GenTree* op1     = node->gtGetOp1();
+    if (op1 == nullptr)
+    {
+        assert((numArgs == 0) && (node->gtGetOp2() == nullptr));
```
nit: It is generally nice to break `&&` asserts into two separate ones. That makes it much clearer which assert is the problematic one when they are hit.

CarolEidt replied:
I generally disagree. Asserts already clutter the code, and though they are useful, they should impact the readability & flow of the code as little as possible, IMO.

mikedn replied:
And this is another bit of code that will go away when I get rid of `GT_LIST`...

tannergooding commented on src/jit/codegenlinear.cpp:
```diff
@@ -1389,6 +1399,55 @@ void CodeGen::genConsumeOperands(GenTreeOp* tree)
+            genConsumeRegs(operand);
+            foundArgs++;
+        }
+        assert(foundArgs == numArgs);
```
Don't we have some intrinsics where `numArgs == -1` since there are multiple overloads?

For example: `SSE_ReciprocalScalar` or `BMI2_MultiplyNoFlags`

CarolEidt replied:
Right. But `HWIntrinsicInfo::lookupNumArgs()` will handle that case and return the correct number.

tannergooding replied:
ah, right.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -109,32 +109,50 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
-                {
-                    assert((ival >= 0) && (ival <= 127));
-                    genHWIntrinsic_R_RM_I(node, ins, (int8_t)ival);
+                    genConsumeAddress(op1);
```
I believe you mentioned that `genConsumeOperands` will also handle the address case, but `genConsumeAddress` is more efficient.

My question then would be: Is it actually worth doing explicit `genConsumeAddress` calls here? It looks like it makes the overall logic slightly more complex...

CarolEidt replied:
In what way does it make the logic more complex? It's probably a matter of taste, but this case needs special handling anyway, and it seems clearest to explicitly consume an address.

tannergooding replied:
>  It's probably a matter of taste

Possibly. I would have thought that the existing `genConsumeOperands(node)` call would have handled things, in which case the changes could have been minimized to just changing `emit->emitIns_R_AR(ins, simdSize, targetReg, op1Reg, 0);` to:
```
// Until we improve the handling of addressing modes in the emitter, we'll create a
// temporary GT_IND to generate code with.
GenTreeIndir load = indirForm(node->TypeGet(), op1);
emit->emitInsLoadInd(ins, simdSize, node->gtRegNum, &load);
```

Likewise, down on L144, just the if block could have been added and the existing `genConsumeOperands(node)` would have handle both the memory store and non-memory store cases (rather than splitting this into 4 separate calls).

mikedn replied:
Please avoid `genConsumeOperands` in SIMD and HWIntrinsic code, it's incompatible with my pending `GT_LIST` changes.

fiigii commented on tests/src/JIT/Regression/JitBlue/GitHub_19550/GitHub_19550.cs:
```diff
@@ -0,0 +1,63 @@
+                // This is the aligned case.
+                Sse2.StoreScalar(p + alignmentOffset + 2, Sse2.Subtract(v, Sse2.LoadAlignedVector128(p + offset + alignmentOffset + 4)));
+                // This is the unaligned case.
+                Sse2.StoreScalar(p + alignmentOffset + 1, Sse2.Subtract(v, Sse2.LoadVector128(p + offset + alignmentOffset + 1)));
```
The test cases look not enough. This PR also changes several `Is4ByteSSEInstruction` points, so I suggest adding a `Sse41.LoadAlignedVector128NonTemporal` test and an `Vector256<T>` load/store test.

In the future, it would be nice to update the intrinsic test template to cover more address modes.

CarolEidt commented:
@redknightlois - I plan to get this in for 3.0, though while waiting for the CI issues to resolve I've gotten side-tracked on other things. I should have an updated PR by the end of this week.

CarolEidt commented:
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0)

CarolEidt commented:
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

fiigii commented on tests/src/JIT/Regression/JitBlue/GitHub_19550/GitHub_19550.cs:
```diff
@@ -25,24 +25,77 @@ struct S
+
+                // Now do a non-temporal load from the aligned location.
+                // That should give us {0, 0, 1, 0}.
+                Vector128<float> v2 = Sse41.LoadAlignedVector128NonTemporal((byte*)(p + alignmentOffset)).AsSingle();
```
`Sse41.IsSupported`?

CarolEidt replied:
Right - of course! Forgot all about the checks!

fiigii commented on tests/src/JIT/Regression/JitBlue/GitHub_19550/GitHub_19550.cs:
```diff
@@ -25,24 +25,77 @@ struct S
+        {
+            try
+            {
+                Avx.Store(p + 1, Avx.Subtract(v, Avx.LoadVector256(p + offset + 1)));
```
`Avx.IsSupported`?

CarolEidt commented:
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

CarolEidt commented:
@dotnet/jit-contrib @tannergooding @fiigii PTAL

fiigii commented:
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitsse2only

@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitsse2only

@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitsse2only
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

fiigii commented on tests/src/JIT/Regression/JitBlue/GitHub_19550/GitHub_19550.cs:
```diff
@@ -0,0 +1,135 @@
+
+                    // Now do a non-temporal load from the aligned location.
+                    // That should give us {0, 0, 1, 0}.
+                    Vector128<float> v2 = Sse41.LoadAlignedVector128NonTemporal((byte*)(p + alignmentOffset)).AsSingle();
```
We need `Sse41.IsSupported` for this case.

CarolEidt replied:
Excellent, thanks. I split this out so that we do an `Sse2` load when `Sse41` isn't available.

CarolEidt commented:
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitsse2only

@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitsse2only

@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitsse2only
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

CarolEidt commented:
@dotnet/jit-contrib @tannergooding ping -  I addressed @fiigii 's feedback on the test case, and fixed an error in the guard for it. I think this is ready to merge.
